VariableMap and functions added.

marisaleung opened the pull request:
This CL does the following:

- Replaces VariableList with VariableMap.

- Edits all workspace.js functions that used VariableList to use VariableMap.

- Adds the functions outlined by the Design Doc: createVariable(name, opt_type, opt_id), getVariable(name), getVariableById(id), deleteVariable(name), deleteVariableById(id), renameVariable(oldName, newName), renameVariableById(id, newName), getVariablesOfType(type), getVariableTypes()

- Tests that the functionality of every edited function works the same as it did prior to editing.

- Tests that the additional functions work at outlined by the Design Doc.

- Adjusts any calls made by files outside of workspace that directly or indirectly accessed variableList.  The calls are adjusted so that original behavior is preserved.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly/1071)
<!-- Reviewable:end -->

RoboErikG commented:
Review status: 0 of 11 files reviewed at latest revision, 17 unresolved discussions.

---

*[core/variables.js, line 91 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4Ay0MwioztZeca7_Z:-Kj4Ay0MwioztZeca7__:b-nlectp) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/variables.js#L91)):*
> ```JavaScript
>                  'Blockly.Variables.allUsedVariables');
>   }
>   return root.getVariablesOfType('');
> ```

allVariables should probably return a list containing every variable in the map. Generators will need this to create the set of initializers for everything. You can do this by iterating through the keys and adding each type's list of variables to the full list before returning it.

---

*[core/variables.js, line 165 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4CHFYFQu7vzOCThqV:-Kj4CHFZox6HHzh_B8Hv:bqqvuk4) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/variables.js#L165)):*
> ```JavaScript
> */
> Blockly.Variables.generateUniqueName = function(workspace) {
>   var variableList = workspace.getVariablesOfType('');
> ```

allVariables. This name should be unique across everything, not just the empty type.

---

*[core/workspace.js, line 84 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4CnvGhmoiPhKUiSde:-Kj4CnvGhmoiPhKUiSdf:b-79ogns) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L84)):*
> ```JavaScript
>    * that are not currently in use.
>    */
>   this.variableMap = {};
> ```

Make this private while you're at it. That way we won't have to break devs again if we ever change this data structure in the future.

---

*[core/workspace.js, line 197 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4DBCr6NQ_H6JWlula:-Kj4DBCs7XEuWTlbktVV:b-spoah9) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L197)):*
> ```JavaScript
>     Blockly.Events.setGroup(false);
>   }
>   for (var key of Object.keys(this.variableMap)) {
> ```

Why do you need to do this much work? You should be able to just clear the map and let gc clean up the individual lists.

---

*[core/workspace.js, line 211 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4KAaYtLlduQO1tZ70:-Kj4KAaYtLlduQO1tZ71:b-qfjiig) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L211)):*
> ```JavaScript
>  * @param {boolean} clearList True if the old variable list should be cleared.
>  */
> Blockly.Workspace.prototype.updateVariableMap = function(clearList) {
> ```

Revert method name change. Let's keep as much of the API as we can.

---

*[core/workspace.js, line 227 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4FU-feOMU6l8ABoRd:-Kj4Dli8_4qC55eEH65H:b2cbx7s) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L227)):*
> ```JavaScript
>     }
>     else {
>       varList.push({'name': variable, 'type': null, 'id': null});
> ```

Please add a TODO to use variable.type and variable.getId() once variable instances are storing more than just name.

---

*[core/workspace.js, line 279 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4G6xDSgXQ5l0NK61c:-Kj4G6xDSgXQ5l0NK61d:bj6qh3v) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L279)):*
> ```JavaScript
>   }
>   var type = (variable || newVariable).type;
>   for (var i = 0, tempvar; tempvar = this.variableMap[type][i]; i++) {
> ```

Instead of iterating through manually if you get the list for just the type of interest and you can use the old code as is. 
var variableList = getVariablesOfType(type);

---

*[core/workspace.js, line 282 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4FImK1Ho3QTKMjrAF:-Kj4FImK1Ho3QTKMjrAG:brvjyqw) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L282)):*
> ```JavaScript
>     // Delete variable_model if newName exists and this is the oldName,
>     // such that the temp variable is not the oldCase or newName.
>     if (variable && tempvar.name == variable.name && newVariable &&
> ```

Add a TODO to update all variable references to point to the new id once they're storing more than just name.

---

*[core/workspace.js, line 302 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4GqWYR7zbRodkZaSG:-Kj4GqWYR7zbRodkZaSH:bmcg7bg) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L302)):*
> ```JavaScript
>  * @param {string} newName New variable name.
>  */
> Blockly.Workspace.prototype.renameVariableById = function(id, newName) {
> ```

renameVariableInstance = function(variable, newName) where variable is an instance of variable including the id, name, and type might make the implementation easier. Fine to use this as a helper if there are places where you have the id without the full variable instance.

---

*[core/workspace.js, line 321 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4HSd7Iv1WwBoPfNwO:-Kj4HSd8VfHunwQJY0dA:b-mduno4) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L321)):*
> ```JavaScript
> Blockly.Workspace.prototype.createVariable = function(name, opt_type, opt_id) {
>   if (this.getVariable(name)) {
>     return;
> ```

You should check that either the type/id aren't set or that they match the existing variable.

---

*[core/workspace.js, line 365 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4HlTdVCwzEfWEmlN_:-Kj4HlTdVCwzEfWEmlNa:bubnapm) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L365)):*
> ```JavaScript
> 
> /**
>  * Delete a variables by the passed in name and all of its uses from this
> ```

remove s on 'variables'

---

*[core/workspace.js, line 419 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4IGTfRoLbdPBdp56y:-Kj4IGTfRoLbdPBdp56z:bkrait) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L419)):*
> ```JavaScript
>  * @private
>  */
> Blockly.Workspace.prototype.deleteVariableInternal_ = function(name) {
> ```

Not in scope for this CL, but in a followup CL you should convert all the internal calls to take a VariableModel or id instead of the name since the name may not correctly identify the variable. Also, if you take the model you only need to check the list with the correct type and can use variableList = getVariablesOfType()

---

*[core/workspace.js, line 431 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4I_jAtNK_KjnJsD2d:-Kj4I_jAtNK_KjnJsD2e:b-diasm4) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L431)):*
> ```JavaScript
>       if (Blockly.Names.equals(variable.name, name)) {
>         delete this.variableMap[key][i];
>         this.variableMap[key].splice(i, 1);
> ```

Once the variable has been found and removed you can exit the loop.

---

*[core/workspace.js, line 622 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4JGWWMVpGqO-qg-Hf:-Kj4JGWWMVpGqO-qg-Hg:b-vvj7n1) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L622)):*
> ```JavaScript
> /**
>  * Find the variable names with the specified type. If type is null, return list
>  *     of all variables.
> ```

if type is null, returns all variables with the empty string type.

---

*[core/workspace.js, line 628 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4JB5i81-184UWgoT1:-Kj4JB5i81-184UWgoT2:b2aldry) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L628)):*
> ```JavaScript
>  */
> Blockly.Workspace.prototype.getVariablesOfType = function(type) {
>   if (!type) {
> ```

Just do type = type || '';

---

*[core/workspace.js, line 635 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4JSpT-UVAALkGqipD:-Kj4JSpT-UVAALkGqipE:bxp3ftn) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L635)):*
> ```JavaScript
>   if (variable_list) {
>     for (var variable of variable_list) {
>       result_list.push(variable.name);
> ```

Prefer you push the VariableModel object. We're going to want to start using it more throughout the code so we can refer to the type and the id.

---

*[core/workspace.js, line 654 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4K0tuRljiC-FIm2Br:-Kj4K0tuRljiC-FIm2Bs:b-il7tuc) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L654)):*
> ```JavaScript
>  */
> Blockly.Workspace.prototype.getAllVariables = function() {
>   var all_variables = [];
> ```

return Blockly.Variables.allVariables(this);

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1071)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
Review status: 0 of 11 files reviewed at latest revision, 20 unresolved discussions.

---

*[core/variables.js, line 225 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4wFqUtfa67Btn9-mE:-Kj4wFqUtfa67Btn9-mF:bju0dua) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/variables.js#L225)):*
> ```JavaScript
>                 });
>           }
>           else if (!Blockly.Procedures.isLegalName_(text, workspace)) {
> ```

What's going on here?  Why do you care about procedures here?

---

*[core/workspace.js, line 84 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4CnvGhmoiPhKUiSde:-Kj4wO7BbKgpozz0aLtU:b32v57) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L84)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Make this private while you're at it. That way we won't have to break devs again if we ever change this data structure in the future.
</blockquote></details>

+1

---

*[core/workspace.js, line 131 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4ye8wlQdTmnryMVQa:-Kj4ye8xDLeGDODBROZk:bvcwv8w) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L131)):*
> ```JavaScript
>   // flyout, so the contents of the dropdown need to be correct.
>   var variables = Blockly.Variables.allUsedVariables(block);
>   for (var variable of variables) {
> ```

for...of is not supported in Internet Explorer: https://docs.microsoft.com/en-us/scripting/javascript/reference/for-dot-dot-dot-of-statement-javascript#requirements

---

*[core/workspace.js, line 211 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4KAaYtLlduQO1tZ70:-Kj4x40vLHfctX2nvYzi:b-3jcdpg) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L211)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Revert method name change. Let's keep as much of the API as we can.
</blockquote></details>

I'm less concerned about preserving the current API (it's not that old) than I am with making sure that the names match what's actually happening in the future.  So I would rather switch to updateVariableMap, or else something generic like updateVariableStore.

If you *are* updating the name, update the parameter `clearList` as well.

---

*[core/workspace.js, line 214 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4z1oVEXQ6W7BPTMmI:-Kj4z1oWL_12MFtiy4wC:b-u068p0) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L214)):*
> ```JavaScript
>   // TODO: Sort
>   if (this.isFlyout) {
>     return
> ```

Missing semicolon.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1071)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
Review status: 0 of 11 files reviewed at latest revision, 24 unresolved discussions.

---

*[tests/jsunit/workspace_test.js, line 27 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj59U8nQDXZBkjxbBkn:-Kj59U8nQDXZBkjxbBko:b-bsnq7b) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/tests/jsunit/workspace_test.js#L27)):*
> ```JavaScript
> var workspace;
> var mockControl_;
> Blockly.Msg.DELETE_VARIABLE = '%1';
> ```

You should change this message in setUp and tearDown, or (even better) only in test cases where it'll be used.  Make sure you reset it at the end, and add a comment explaining why you need to define this (which function uses it, and why it's set to this particular value).

---

*[tests/jsunit/workspace_test.js, line 79 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj51pwoaVfTjQL3JmKT:-Kj51pwoaVfTjQL3JmKU:b-kqsz0z) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/tests/jsunit/workspace_test.js#L79)):*
> ```JavaScript
>  */
> function checkVariableValues(name, type, id) {
>   assertNotUndefined(variable = workspace.getVariable(name));
> ```

I prefer to keep assignments out of asserts whenever possible.  I would do
```
var variable = workspace.getVariable(name);
assertNotUndefined(variable);
```

---

*[tests/jsunit/workspace_test.js, line 82 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj53Oz1_Zwn9q3h8-l5:-Kj53Oz2Dx50dbkJ4hLy:b-24mcfw) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/tests/jsunit/workspace_test.js#L82)):*
> ```JavaScript
>   assertEquals(name, variable.name);
>   assertEquals(type, variable.type);
>   assertEquals(id, variable.getId());
> ```

Why do you access variable.name and variable.type directly, but the id though variable.getId()?

---

*[tests/jsunit/workspace_test.js, line 145 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj54hkkGWEKyx8YL1Bq:-Kj54hkkGWEKyx8YL1Br:b-jzphtv) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/tests/jsunit/workspace_test.js#L145)):*
> ```JavaScript
>     blockB && blockB.dispose();
>     blockA && blockA.dispose();
>     workspaceTest_tearDown();
> ```

You may no longer need to dispose of the blocks manually, as they may be disposed of by workspace dispose.  In that case you would also not need to declare them outside of the try.

Same applies to all of these tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1071)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
Review status: 0 of 11 files reviewed at latest revision, 24 unresolved discussions.

---

*[tests/jsunit/workspace_test.js, line 82 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj53Oz1_Zwn9q3h8-l5:-Kj5ASq1PfDSbvoq8Cw6:b-ggpojs) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/tests/jsunit/workspace_test.js#L82)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Why do you access variable.name and variable.type directly, but the id though variable.getId()?
</blockquote></details>

It's because the ID is read only, so it has an accessor but no setter.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1071)*
<!-- Sent from Reviewable.io -->

marisaleung commented:
Review status: 0 of 11 files reviewed at latest revision, 24 unresolved discussions.

---

*[core/variables.js, line 91 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4Ay0MwioztZeca7_Z:-Kj4zLlNbVlYxcxEwoml:b-896fix) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/variables.js#L91)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

allVariables should probably return a list containing every variable in the map. Generators will need this to create the set of initializers for everything. You can do this by iterating through the keys and adding each type's list of variables to the full list before returning it.
</blockquote></details>

Done.

---

*[core/variables.js, line 165 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4CHFYFQu7vzOCThqV:-Kj4zjFr5YSKm0zBrAqA:b-896fix) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/variables.js#L165)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

allVariables. This name should be unique across everything, not just the empty type.
</blockquote></details>

Done.

---

*[core/variables.js, line 225 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4wFqUtfa67Btn9-mE:-Kj594HSY1zXNEyiBDPg:b-ki53ab) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/variables.js#L225)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

What's going on here?  Why do you care about procedures here?
</blockquote></details>

Following Design Doc and checking that name is unique across all variables and procedures. If you want me to move that to a different CL, let me know.

---

*[core/workspace.js, line 84 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4CnvGhmoiPhKUiSde:-Kj4zrfSBukHGLa9eE_Q:b-896fix) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L84)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Make this private while you're at it. That way we won't have to break devs again if we ever change this data structure in the future.
</blockquote></details>

Done.

---

*[core/workspace.js, line 131 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4ye8wlQdTmnryMVQa:-Kj5_0pPaK6RzU-R7Sj4:b-wpvnxc) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L131)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

for...of is not supported in Internet Explorer: https://docs.microsoft.com/en-us/scripting/javascript/reference/for-dot-dot-dot-of-statement-javascript#requirements
</blockquote></details>

Discussed offline to use a format like this:
var keys = Object.keys(variableMap);
for (var i = 0; i < keys.length; i++) {

Changed each for loop to not use for...of

Done.

---

*[core/workspace.js, line 197 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4DBCr6NQ_H6JWlula:-Kj5-TzdFaFxwwEbZMx2:b60r3sb) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L197)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Why do you need to do this much work? You should be able to just clear the map and let gc clean up the individual lists.
</blockquote></details>

Hm I thought I had been running into a problem in the tests where garbage collection wasn't working. Anywho, this.variableMap_ = {} currently passes all the tests so I'll use that.

---

*[core/workspace.js, line 211 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4KAaYtLlduQO1tZ70:-Kj50b7sP0GqHbdNcBce:b71kbk8) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L211)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

I'm less concerned about preserving the current API (it's not that old) than I am with making sure that the names match what's actually happening in the future.  So I would rather switch to updateVariableMap, or else something generic like updateVariableStore.

If you *are* updating the name, update the parameter `clearList` as well.
</blockquote></details>

I change it to updateVariableStore and changed clearList to clear. Let me know if you want me to revert or do something else.

---

*[core/workspace.js, line 214 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4z1oVEXQ6W7BPTMmI:-Kj50pNoCGuncTSFhzsw:b-896fix) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L214)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Missing semicolon.
</blockquote></details>

Done.

---

*[core/workspace.js, line 227 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4FU-feOMU6l8ABoRd:-Kj516FQ3sWBTk2zTb2Y:b-896fix) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L227)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Please add a TODO to use variable.type and variable.getId() once variable instances are storing more than just name.
</blockquote></details>

Done.

---

*[core/workspace.js, line 279 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4G6xDSgXQ5l0NK61c:-Kj51CTAZnHS98w1l1jQ:b2b1ka6) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L279)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Instead of iterating through manually if you get the list for just the type of interest and you can use the old code as is. 
var variableList = getVariablesOfType(type);
</blockquote></details>

Yay reusing code! Thank you for the reminder. Done.

---

*[core/workspace.js, line 282 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4FImK1Ho3QTKMjrAF:-Kj56WjRB2rdRYdOC_5w:b-896fix) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L282)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Add a TODO to update all variable references to point to the new id once they're storing more than just name.
</blockquote></details>

Done.

---

*[core/workspace.js, line 302 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4GqWYR7zbRodkZaSG:-Kj6PpfB0RZR9bAu4Y2g:b-896fix) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L302)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

renameVariableInstance = function(variable, newName) where variable is an instance of variable including the id, name, and type might make the implementation easier. Fine to use this as a helper if there are places where you have the id without the full variable instance.
</blockquote></details>

Done.

---

*[core/workspace.js, line 321 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4HSd7Iv1WwBoPfNwO:-Kj56zi4Q-NZ2p8otDju:bk65lr5) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L321)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

You should check that either the type/id aren't set or that they match the existing variable.
</blockquote></details>

1. If the variable does exist and the type/id isn't set, do I update them?
2. If they don't match the existing variable, do I update them or throw an error?

---

*[core/workspace.js, line 365 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4HlTdVCwzEfWEmlN_:-Kj57JjOZRFvwvGKkCvq:b-896fix) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L365)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

remove s on 'variables'
</blockquote></details>

Done.

---

*[core/workspace.js, line 431 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4I_jAtNK_KjnJsD2d:-Kj58OhyTcLzN_svo46g:b-896fix) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L431)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Once the variable has been found and removed you can exit the loop.
</blockquote></details>

Done.

---

*[core/workspace.js, line 622 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4JGWWMVpGqO-qg-Hf:-Kj58XLCUyMZwcfSbdmN:bly8s6w) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L622)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

if type is null, returns all variables with the empty string type.
</blockquote></details>

Done. Function Description updated too.

---

*[core/workspace.js, line 628 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4JB5i81-184UWgoT1:-Kj58c1X495CUF6W4BLi:b-896fix) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L628)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Just do type = type || '';
</blockquote></details>

Done.

---

*[core/workspace.js, line 635 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4JSpT-UVAALkGqipD:-Kj58d6-I7sYAoDWbgcb:b-896fix) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L635)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Prefer you push the VariableModel object. We're going to want to start using it more throughout the code so we can refer to the type and the id.
</blockquote></details>

Done.

---

*[core/workspace.js, line 654 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4K0tuRljiC-FIm2Br:-Kj58pBVSDwceBm911q_:b6rolx8) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L654)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

return Blockly.Variables.allVariables(this);
</blockquote></details>

Actually, I have Blockly.Variables.allVariables() call this function instead of the other way around. I think it makes sense to keep the actual searching in workspace.js because variableMap_ is there.  What do you think?

---

*[tests/jsunit/workspace_test.js, line 27 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj59U8nQDXZBkjxbBkn:-Kj6OKrlRcWRGGVNRzia:b-896fix) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/tests/jsunit/workspace_test.js#L27)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

You should change this message in setUp and tearDown, or (even better) only in test cases where it'll be used.  Make sure you reset it at the end, and add a comment explaining why you need to define this (which function uses it, and why it's set to this particular value).
</blockquote></details>

Done.

---

*[tests/jsunit/workspace_test.js, line 79 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj51pwoaVfTjQL3JmKT:-Kj6DyMrUwyOt_2v9cvT:bcsf63k) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/tests/jsunit/workspace_test.js#L79)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

I prefer to keep assignments out of asserts whenever possible.  I would do
```
var variable = workspace.getVariable(name);
assertNotUndefined(variable);
```
</blockquote></details>

Done. And fixed throughout the rest of the tests I created.

---

*[tests/jsunit/workspace_test.js, line 82 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj53Oz1_Zwn9q3h8-l5:-Kj5_XWYhaAsKaLuMUdN:b-896fix) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/tests/jsunit/workspace_test.js#L82)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

It's because the ID is read only, so it has an accessor but no setter.
</blockquote></details>

Done.

---

*[tests/jsunit/workspace_test.js, line 145 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj54hkkGWEKyx8YL1Bq:-Kj6DkJxzGODrDUroQFZ:b-896fix) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/tests/jsunit/workspace_test.js#L145)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

You may no longer need to dispose of the blocks manually, as they may be disposed of by workspace dispose.  In that case you would also not need to declare them outside of the try.

Same applies to all of these tests.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1071)*
<!-- Sent from Reviewable.io -->

marisaleung commented:
Review status: 0 of 11 files reviewed at latest revision, 24 unresolved discussions.

---

*[core/workspace.js, line 84 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4CnvGhmoiPhKUiSde:-Kj6Pw5QnxDE9iMWkX45:b-896fix) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L84)):*
<details><summary><i>Previously, marisaleung wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---

*[core/workspace.js, line 419 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4IGTfRoLbdPBdp56y:-Kj6PxFWsosMUyLodhMU:b-54sncc) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L419)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

Not in scope for this CL, but in a followup CL you should convert all the internal calls to take a VariableModel or id instead of the name since the name may not correctly identify the variable. Also, if you take the model you only need to check the list with the correct type and can use variableList = getVariablesOfType()
</blockquote></details>

Acknowledged!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1071)*
<!-- Sent from Reviewable.io -->

marisaleung commented:
Review status: 0 of 11 files reviewed at latest revision, 22 unresolved discussions.

---

*[core/workspace.js, line 321 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4HSd7Iv1WwBoPfNwO:-KjAD1I2WrWNYKTMKRgV:b-54sjha) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/workspace.js#L321)):*
<details><summary><i>Previously, marisaleung wrote…</i></summary><blockquote>

1. If the variable does exist and the type/id isn't set, do I update them?
2. If they don't match the existing variable, do I update them or throw an error?

</blockquote></details>

Offline we decided:
If Type and Id are defined in parameters and if there is a variable that already exists that don't match, throw an error.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1071)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
Reviewed 1 of 11 files at r1, 1 of 4 files at r2.
Review status: 2 of 11 files reviewed at latest revision, 28 unresolved discussions.

---

*[core/variables.js, line 225 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4wFqUtfa67Btn9-mE:-KjAFKa_M3tv0w4cc6S0:b1qa34q) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/variables.js#L225)):*
<details><summary><i>Previously, marisaleung wrote…</i></summary><blockquote>

Following Design Doc and checking that name is unique across all variables and procedures. If you want me to move that to a different CL, let me know.
</blockquote></details>

Where is that specified in the design doc?  I don't think that's behaviour that we want.

---

*[core/variables.js, line 89 at r2](https://reviewable.io:443/reviews/google/blockly/1071#-KjAEmGJmeWavjzW9gn4:-KjAEmGJmeWavjzW9gn5:b-yizjn7) ([raw file](https://github.com/google/blockly/blob/2994805e6ac2d3edf14283874aa14ba8666efff5/core/variables.js#L89)):*
> ```JavaScript
>     console.warn('Deprecated call to Blockly.Variables.allVariables ' +
>                  'with a block instead of a workspace.  You may want ' +
>                  'Blockly.Variables.allUsedVariables');
> ```

Return early if it's a block, or else get the workspace from the block and update the message to say that you're making a best guess at the workspace.

This was a preexisting bug, but now it'll break instead of returning undefined.

---

*[core/workspace.js, line 132 at r2](https://reviewable.io:443/reviews/google/blockly/1071#-KjAGtRe8tj4lYq1TtLj:-KjAGtRe8tj4lYq1TtLk:b-41u46f) ([raw file](https://github.com/google/blockly/blob/2994805e6ac2d3edf14283874aa14ba8666efff5/core/workspace.js#L132)):*
> ```JavaScript
>   // flyout, so the contents of the dropdown need to be correct.
>   var variables = Blockly.Variables.allUsedVariables(block);
>   for (var i = 0, variable; variable = variables[i]; i++) {
> ```

Since there's now a distinction between a variable name (a string) and a variable (a variableModel), use "name" instead of "variable" here.  That makes the following code
```
if (!this.getVariable(name)) {
  this.createVariable(name);
}
```
which reads more cleanly.

---

*[core/workspace.js, line 211 at r2](https://reviewable.io:443/reviews/google/blockly/1071#-KjAGeSY5ausj8hnfw2k:-KjAGeSY5ausj8hnfw2l:bt6aeke) ([raw file](https://github.com/google/blockly/blob/2994805e6ac2d3edf14283874aa14ba8666efff5/core/workspace.js#L211)):*
> ```JavaScript
>     return;
>   }
>   // Update the list in place so that the flyout's references stay correct.
> ```

This comment belongs closer to where the update is happening.

---

*[core/workspace.js, line 214 at r2](https://reviewable.io:443/reviews/google/blockly/1071#-KjAH6iIAhix3AzT7rDD:-KjAH6iIAhix3AzT7rDE:b-t4128) ([raw file](https://github.com/google/blockly/blob/2994805e6ac2d3edf14283874aa14ba8666efff5/core/workspace.js#L214)):*
> ```JavaScript
>   var allVariables = Blockly.Variables.allUsedVariables(this);
>   var varList = [];
>   for (var i = 0, variable; variable = allVariables[i]; i++) {
> ```

As before, `name` instead of `variable`

---

*[core/workspace.js, line 256 at r2](https://reviewable.io:443/reviews/google/blockly/1071#-KjAHhAkOmzPGIKtbYs7:-KjAHhAkOmzPGIKtbYs8:beop7pi) ([raw file](https://github.com/google/blockly/blob/2994805e6ac2d3edf14283874aa14ba8666efff5/core/workspace.js#L256)):*
> ```JavaScript
>          '". Both must be the same type.');
>   }
>   var type = (variable || newVariable || '').type;
> ```

What is `''.type`?

---

*[core/workspace.js, line 259 at r2](https://reviewable.io:443/reviews/google/blockly/1071#-KjAI9AkqSKnE8TTZ0XU:-KjAI9AkqSKnE8TTZ0XV:b-7vpy7f) ([raw file](https://github.com/google/blockly/blob/2994805e6ac2d3edf14283874aa14ba8666efff5/core/workspace.js#L259)):*
> ```JavaScript
>   var variableList = this.getVariablesOfType(type);
>   if (variable) {
>     variableIndex = variableList.indexOf(variable);
> ```

When would `variable` be falsy?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1071)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
Review status: 2 of 11 files reviewed at latest revision, 29 unresolved discussions.

---

*[core/workspace.js, line 289 at r2](https://reviewable.io:443/reviews/google/blockly/1071#-KjAPM1PfJthFXRr07Sn:-KjAPM1PfJthFXRr07So:b-qz3oqm) ([raw file](https://github.com/google/blockly/blob/2994805e6ac2d3edf14283874aa14ba8666efff5/core/workspace.js#L289)):*
> ```JavaScript
>     // Only changing case, or renaming to a completely novel name.
>     if (oldCase) {
>       this.variableMap_[type][newVariableIndex].name = newName;
> ```

You can remove this case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1071)*
<!-- Sent from Reviewable.io -->

marisaleung commented:
Review status: 2 of 11 files reviewed at latest revision, 29 unresolved discussions.

---

*[core/variables.js, line 225 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4wFqUtfa67Btn9-mE:-KjAQa77Chg-psh3Hv6-:b-zcdd22) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/variables.js#L225)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Where is that specified in the design doc?  I don't think that's behaviour that we want.
</blockquote></details>

https://docs.google.com/document/d/1ZCLus7iBrL7ZMhpmlCdrVDKXEpQtwqWdDjf_Ojm88yE/edit#heading=h.fcgc975dvzod

In the description of this.name

---

*[core/variables.js, line 89 at r2](https://reviewable.io:443/reviews/google/blockly/1071#-KjAEmGJmeWavjzW9gn4:-KjARS9gSTVVuE0g4Vcu:b-896fix) ([raw file](https://github.com/google/blockly/blob/2994805e6ac2d3edf14283874aa14ba8666efff5/core/variables.js#L89)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Return early if it's a block, or else get the workspace from the block and update the message to say that you're making a best guess at the workspace.

This was a preexisting bug, but now it'll break instead of returning undefined.
</blockquote></details>

Done.

---

*[core/workspace.js, line 132 at r2](https://reviewable.io:443/reviews/google/blockly/1071#-KjAGtRe8tj4lYq1TtLj:-KjAUfHRrOksVJXktwrO:b-896fix) ([raw file](https://github.com/google/blockly/blob/2994805e6ac2d3edf14283874aa14ba8666efff5/core/workspace.js#L132)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Since there's now a distinction between a variable name (a string) and a variable (a variableModel), use "name" instead of "variable" here.  That makes the following code
```
if (!this.getVariable(name)) {
  this.createVariable(name);
}
```
which reads more cleanly.
</blockquote></details>

Done.

---

*[core/workspace.js, line 211 at r2](https://reviewable.io:443/reviews/google/blockly/1071#-KjAGeSY5ausj8hnfw2k:-KjARlNaWZ5tB4bViziJ:b-896fix) ([raw file](https://github.com/google/blockly/blob/2994805e6ac2d3edf14283874aa14ba8666efff5/core/workspace.js#L211)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

This comment belongs closer to where the update is happening.
</blockquote></details>

Done.

---

*[core/workspace.js, line 214 at r2](https://reviewable.io:443/reviews/google/blockly/1071#-KjAH6iIAhix3AzT7rDD:-KjAS_-YX95vlVL0Xchn:b-896fix) ([raw file](https://github.com/google/blockly/blob/2994805e6ac2d3edf14283874aa14ba8666efff5/core/workspace.js#L214)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

As before, `name` instead of `variable`
</blockquote></details>

Done.

---

*[core/workspace.js, line 256 at r2](https://reviewable.io:443/reviews/google/blockly/1071#-KjAHhAkOmzPGIKtbYs7:-KjAU4cpf2p1CkGqbkg6:b-kclilb) ([raw file](https://github.com/google/blockly/blob/2994805e6ac2d3edf14283874aa14ba8666efff5/core/workspace.js#L256)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

What is `''.type`?
</blockquote></details>

It is undefined which still works because creating a variable with an undefined type still sets the type to ''. But to make it more clear, I added an if else statement.

---

*[core/workspace.js, line 259 at r2](https://reviewable.io:443/reviews/google/blockly/1071#-KjAI9AkqSKnE8TTZ0XU:-KjAUU8P5ajzj-AcMwIS:b-3l9jun) ([raw file](https://github.com/google/blockly/blob/2994805e6ac2d3edf14283874aa14ba8666efff5/core/workspace.js#L259)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

When would `variable` be falsy?
</blockquote></details>

If we are trying to rename a variable that does not exist.

---

*[core/workspace.js, line 289 at r2](https://reviewable.io:443/reviews/google/blockly/1071#-KjAPM1PfJthFXRr07Sn:-KjAUY10TIQ_RDbGZtTy:b-896fix) ([raw file](https://github.com/google/blockly/blob/2994805e6ac2d3edf14283874aa14ba8666efff5/core/workspace.js#L289)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

You can remove this case.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1071)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
Review status: 1 of 11 files reviewed at latest revision, 21 unresolved discussions.

---

*[core/variables.js, line 225 at r1](https://reviewable.io:443/reviews/google/blockly/1071#-Kj4wFqUtfa67Btn9-mE:-KjAZaAEcRut3lRogrzX:b-n9pny3) ([raw file](https://github.com/google/blockly/blob/6c1a5b2bef9f3dd2c613744a4be0ad54a9623f6a/core/variables.js#L225)):*
<details><summary><i>Previously, marisaleung wrote…</i></summary><blockquote>

https://docs.google.com/document/d/1ZCLus7iBrL7ZMhpmlCdrVDKXEpQtwqWdDjf_Ojm88yE/edit#heading=h.fcgc975dvzod

In the description of this.name
</blockquote></details>

Interesting.  I'd like to discuss that, but I'll move it to the doc.  Don't worry about it on this PR.

---

*[core/variables.js, line 89 at r2](https://reviewable.io:443/reviews/google/blockly/1071#-KjAEmGJmeWavjzW9gn4:-KjAYYAlChIGy1FJI_-5:b-be2j1g) ([raw file](https://github.com/google/blockly/blob/2994805e6ac2d3edf14283874aa14ba8666efff5/core/variables.js#L89)):*
<details><summary><i>Previously, marisaleung wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Return an empty array so that you don't break calling code.

---

*[core/workspace.js, line 85 at r2](https://reviewable.io:443/reviews/google/blockly/1071#-KjA_8iuvKoQ0fc2EBDV:-KjA_8iuvKoQ0fc2EBDW:b-4rxn6o) ([raw file](https://github.com/google/blockly/blob/2994805e6ac2d3edf14283874aa14ba8666efff5/core/workspace.js#L85)):*
> ```JavaScript
>    * @private
>    */
>   this.variableMap_ = {};
> ```

Object.create(null)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1071)*
<!-- Sent from Reviewable.io -->

marisaleung commented:
Review status: 1 of 11 files reviewed at latest revision, 21 unresolved discussions.

---

*[core/variables.js, line 89 at r2](https://reviewable.io:443/reviews/google/blockly/1071#-KjAEmGJmeWavjzW9gn4:-KjEQoPS7r3Mf6BdxpgR:b-896fix) ([raw file](https://github.com/google/blockly/blob/2994805e6ac2d3edf14283874aa14ba8666efff5/core/variables.js#L89)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Return an empty array so that you don't break calling code.
</blockquote></details>

Done.

---

*[core/workspace.js, line 85 at r2](https://reviewable.io:443/reviews/google/blockly/1071#-KjA_8iuvKoQ0fc2EBDV:-KjEPqBPTFK0FlMQCXiV:bfqcven) ([raw file](https://github.com/google/blockly/blob/2994805e6ac2d3edf14283874aa14ba8666efff5/core/workspace.js#L85)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Object.create(null)
</blockquote></details>

Originally I was following javascript style guide rule:
https://engdoc.corp.google.com/eng/doc/devguide/js/styleguide.shtml?cl=head#features-objects-ctor

But I found this thread that OK's the use of Object.create(null). So this shall be done!

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1071)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r4.
Review status: 2 of 11 files reviewed at latest revision, 21 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1071#-:-KjERgaaGm1JCY6OXJJ8:bnfp4nl)*
<!-- Sent from Reviewable.io -->
